### ISSUE TITLE
Fail build if `nvm install` fails

### ISSUE THREAD

**Issue by Turbo87** [travis-ci/travis-ci#7001]
The `nvm install` command at the start is currently failing due to [network issues with nodejs.org](https://github.com/nodejs/build/issues/562#issuecomment-266024258), but instead of this failing/erroring the build, it just continues and uses the default Node 0.10, which leads to errors. It would be much better if the build would just fail instead.

**Comment by scasagrande-d2l** [comment]
Yeah, been having strange behaviour with `node: 6` today. One build it decided to try to compile node `6.9.2` from source (32min job!), another run it tried to download for a few minutes and then failed defaulting to `4.1.2` (15min job!), and another run I was getting checksum errors. The last one completed in our normal time period (~2min), but was running on `4.1.2`.

**Comment by BanzaiMan** [comment]
Do you have a build log URL?

I thought we had a mechanism to fail, but the logic must be faulty.

**Comment by Turbo87** [comment]
@BanzaiMan one example of this is https://travis-ci.org/ember-cli/ember-cli-chai/builds/182609593

**Comment by scasagrande-d2l** [comment]
I sure do, let me get that for you:

https://travis-ci.com/Brightspace/bulk-exemptions/builds/36384044
https://travis-ci.com/Brightspace/bulk-exemptions/builds/36384346
https://travis-ci.com/Brightspace/bulk-exemptions/builds/36389925

**Comment by BanzaiMan** [comment]
See also https://github.com/nodejs/build/issues/562 for slow connection to nodejs.org.

**Comment by BanzaiMan** [comment]
The relevant part is https://github.com/travis-ci/travis-build/blob/0501eac99c42b9ed48d2d08e0e7a3b13c021bbc0/lib/travis/build/script/node_js.rb#L136-L145

This compiles to a bash snippet like this:

```sh
travis_cmd nvm\ install\ 4 --echo --timing

if [[ $? -ne 0 ]]; then
  echo -e "\033[31;1mFailed to install 4. Remote repository may not be reachable.\033[0m"
  echo -e "Using locally available version 4, if applicable."
  travis_cmd nvm\ use\ 4 --echo
  if [[ $? -ne 0 ]]; then
    echo -e "\033[31;1mUnable to use 4\033[0m"
    travis_cmd false --assert
  fi
fi
```

Perhaps `nvm` is exiting with status 0, though the installation failed.

**Comment by BanzaiMan** [comment]
Confirmed the above suspicion.

```sh-session
$ nvm install 4
VERSION_PATH=''
#                                                                          1.8%
curl: (56) SSLRead() return error -9806
Binary download from https://nodejs.org/dist/v4.7.0/node-v4.7.0-darwin-x64.tar.xz failed, trying source.
grep: /Users/asari/.nvm/.cache/bin/node-v4.7.0-darwin-x64/node-v4.7.0-darwin-x64.tar.xz: No such file or directory
Provided file to checksum does not exist.
Binary download failed, trying source.
Detected that you have 4 CPU thread(s)
Set the number of jobs to 4 - 1 = 3 jobs to speed up the build
######################################################################## 100.0%
Computing checksum with shasum -a 256
Checksums do not match: '069d5ee4f79eace011758601f9c048f5c22b9673759beaf31308c43842f4f838' found, 'b03c777ba8817a8478d57f00797db86dc7e7953d2066c34edbceeba8ad056142' expected.
tar: Unrecognized archive format
tar: Error exit delayed from previous errors.
nvm: install v4.7.0 failed!
$ echo $?
0
```

**Comment by Turbo87** [comment]
ouch... @BanzaiMan I guess we should open an issue at https://github.com/creationix/nvm/ then

**Comment by gneatgeek** [comment]
I'm also experiencing this issue: https://travis-ci.org/gneatgeek/reveal-leap-motion/builds/182632983

**Comment by alallier** [comment]
I am experience this also https://api.travis-ci.org/jobs/182575815/log.txt?deansi=true

**Comment by derhuerst** [comment]
@gneatgeek @alallier See nodejs/build#562 for the original issue and creationix/nvm#1347 for the follow-up issue in nvm.

**Comment by BanzaiMan** [comment]
Currently, we are blocked by the nvm issue I linked above. We *could* look at the command output to sidestep the issue, but I think this is a bad idea.

**Comment by apepper** [comment]
> Currently, we are blocked by the nvm issue I linked above. We could look at the command output to sidestep the issue, but I think this is a bad idea.

The nvm issue is now resolved. See https://github.com/nodejs/build/issues/562#issuecomment-266089304.

I suggest two improvements to reduce the impact of a similar situation, wich may arise in the future:

1. Adjust `nvm install`, so that it fails the build, if it is not successful.
2. If possible, provide a Travis cache for node versions (e.g. mirror https://nodejs.org/dist/). AFAIKS there are already similar caches for ruby versions. Their has already been some discussions about mirroring nodejs.org (see https://github.com/nodejs/build/issues/562#issuecomment-266050205).

**Comment by BanzaiMan** [comment]
@apepper 
1. The blocking issue is https://github.com/creationix/nvm/issues/1347. https://github.com/nodejs/build/issues/562 is a separate issue which just exposed this problem.
1. "Travis cache" still leaves the build process vulnerable to network issues. No matter where you put the language archives, if they are not available on the build image already, network could be a source of concern. We try to mitigate the possibility by choosing most reliable sources, but the reality is there.

**Comment by PeterDaveHello** [comment]
https://github.com/creationix/nvm/pull/1348 merged, waiting for `nvm` new release, cc @ljharb

**Comment by BanzaiMan** [comment]
There is a new `nvm` release [0.33.2](https://github.com/creationix/nvm/releases/tag/v0.33.2) with the fix, and we have deployed this.
